### ISSUE TITLE
Meta issue: roadmap for exporting output from napari

### ISSUE THREAD

**Issue by GenevieveBuckley** [napari/napari#1068]
## 🚀 Feature

Meta-issue for exporting data, annotations, etc. from napari in a consistent way.

## Motivation

The motivation here is pretty clear: in order for napari to be useful as part of an analysis pipeline it should support basic export capabilities. Users are already asking for these features: https://github.com/napari/napari/issues/382

## Discussion

**I want to hear from everyone about what export formats should be supported.**
I'll update the list below as people comment. 

At this stage, I'd like an emphasis on formats that should be easy for non-programmers to work with,
and formats that give us the greatest compatibility with other tools (eg: for further analysis).

Also, now that @tlambert03 has put in a bunch of work to enable plugins https://github.com/napari/napari/pull/937 (& other PRs),
I'd like to hear from him about how we can best use that infrastructure here as well.

### napari viewer
* serializable, round trip capable

### layer group
* layer groups don't exist yet, but would we also want them to be serializable too?

### image layer
* all imageio formats
* zarr

### labels layer
* all imageio formats
* zarr

### points layer
* csv table
* ImageJ compatible point ROIs
* SVG

### shapes layer
* csv table
* ImageJ compatible ROIs
* SVG
* geoJSON

### surface layer
* PLY
* STL
* OBJ (ASCII variant)

### vector layer
* csv table
* SVG

## Additional context

@danielballan has just started a discussion around a SciPy wide reader protocol
* https://github.com/napari/napari/issues/1066
* https://blog.danallan.com/posts/2020-03-07-reader-protocol/

This PR https://github.com/napari/napari/pull/941, intended to address https://github.com/napari/napari/issues/382 was made before the io plugin infrastructure was in place.
I expect that the `write_csv` function introduced here should probably become a writer plugin instead.

**Comment by sofroniewn** [comment]
@GenevieveBuckley thanks so much for kicking off this issue with all those great examples. One thing I'd like to discuss (could be here or in an even more focused issue) is what the API for our plugin writer hookspec should be.

For me a guiding principle with the plugin writer should be symmetry with the plugin reader. So while the reader ultimately provides a function that takes a path or list of paths and returns an object (list of layer data tuples) I think the writer function should take a path or list of paths, that same object (the list of layer data tuples), and an optional format / extension.

If we change the object the reader returns we should change the object the writer accepts. There should always be symmetry.

How does this sound @GenevieveBuckley @tlambert03 @jni?

If we add a general purpose plugin writers like this then we could meet many of the needs @GenevieveBuckley calls out above.

**Comment by jni** [comment]
@sofroniewn I like the sound of that!

**Comment by GenevieveBuckley** [comment]
I agree, symmetry is a really important goal

**Comment by tlambert03** [comment]
definitely on board with a symmetric approach... though I anticipate some challenging details when it comes to saving _multiple_ layers.  some questions:

- assuming a strictly symmetric approach, (i.e. the user provides a save `path` and `napari` then hands `LayerData` tuples to writer plugins), do we hand the _full_ viewer state (a list of all `LayerData` tuples) to each plugin, or do we iterate over each layer, looking for a writer plugin?
- if we hand each plugin the full list, are they expected to save the full state, and does that mean that every plugin must know how to write _every_ type of layer?  or do they get to pick and choose the ones they know how to write?
- if picking/choosing is allowed and/or if we hand one layer at a time to plugins, how do they get "combined" into the output?
- what's the significance of the output path provided by the user?  `skimage.io.imsave` and other similar functions have that lovely property of using the output filename extension to determine the image format... but is there any single filetype that would work for the full viewer?  Or is it always going to have to be a zip file or directory, and would extensions only be useful on a single-layer basis?
- should also consider multi-channel layers (and eventually layer groups).

we could perhaps have writer hooks for each type of layer: `napari_save_image`, `napari_save_points`, etc... where a hookspec might look like:
```python
@hook_specification
def napari_save_image(path: str, data: ArrayLike, meta: Dict[str, Any]) -> None:
    """Save data and metadata from an image layer."""
```
and then plugins could implement as few or as many layer types as they'd like.  I guess the output in this case would always be a directory (or some sort of archive) and each layer could be handled by one plugin.

I do think we need to allow plugins like this, but I find myself thinking it's less natural than reader plugins.  Alternatively, I can imagine us having one "official" default napari export format that saves the full viewer state, while also allowing plugins to export individual layers when the user selects one or more layers and chooses "save as..." or something like that.  (in which case, perhaps the writer spec needs to declare which layers and layer combinations it can support?)

do those questions bring up any more more ideas or opinions?

**Comment by GenevieveBuckley** [comment]
> I can imagine us having one "official" default napari export format that saves the full viewer state, while also allowing plugins to export individual layers when the user selects one or more layers and chooses "save as..." or something like that. (in which case, perhaps the writer spec needs to declare which layers and layer combinations it can support?)

I'd lean towards this, personally. Mostly because I'm worried about introducing too much complexity early on, but if that's less of an issue then you can take this opinion with a grain of salt.

**Comment by tlambert03** [comment]
I think implementing hooks for plugins to save individual layers will be rather straightforward... but saving _multiple_ arbitrary layers either demands quite a lot from plugin developers, or rather complex internal logic for combining the outputs of disparate plugins.  if we start by implementing our own full-viewer export process, we might realize some more hidden issues in the process.

**Comment by sofroniewn** [comment]
Great comments @tlambert03 @GenevieveBuckley. Just to set terminology I see the following options for saving

1. A single layer
2. An arbitrary list of `LayerData` objects, that may contain data from layers of different types
3. Full viewer state (see #851 for a WIP on viewer serialization), including dims, camera etc.

Right now reading returns `2.` and I wrote above it could be natural to start with writing doing `2.` as well.

I just want to make sure @tlambert03 and I are clear on `2.` vs `3.` as you occasionally wrote "full-viewer" but it sounds like you described `2.`. Maybe I'll further add though that we could be trying to unify `2.` and `3.` at least a little more here https://github.com/napari/napari/issues/1077#issuecomment-602764596.

If we go with `2.` does it make sense to also have writers plugins provide a simple checker that mirrors the reader checker? A writer plugin could for example have a very fast checker that looked at the provided extension say `tif`, then checked that the `LayerData` list had length 1 (or length matching the length of the list of provided paths), and then checked the the `layer_type` was image. If all these things passed then the writer function would be returned and the file would get written. If not, we'd be on to checking the next one.

From the UI we'd have one option that was `save all layers` and that would save all of them, and then we could have another that was `save selected layers` and it would save only the selected ones. If you only wanted one layer saved you'd have to select it and then use that option.

How does this sound to you @tlambert03 @GenevieveBuckley @jni?

**Comment by jni** [comment]
Erm, it seems to me that layers groups solve all our problems... ;)

**Comment by sofroniewn** [comment]
> If we go with 2. does it make sense to also have writers plugins provide a simple checker that mirrors the reader checker? A writer plugin could for example have a very fast checker that looked at the provided extension say tif, then checked that the LayerData list had length 1 (or length matching the length of the list of provided paths), and then checked the the layer_type was image. If all these things passed then the writer function would be returned and the file would get written. If not, we'd be on to checking the next one.

@tlambert03 and others - I'm going to start working on a PR for this now. I'm very happy to change API / approach as I go, but I think it will be good for me to understand how to write a hookspec etc to aid with maintenance of this part of the codebase too

**Comment by tlambert03** [comment]
yeah, that's great.  feel free to slack, zulip, and/or comment here if you want feedback

**Comment by sumanthratna** [comment]
I think it would help to allow Points and Shapes to be exported as SVGs. Of course, this can probably be done programmatically from the CSV output, but it'd be nice to be able to get an SVG without having to write extra code.

**Comment by sofroniewn** [comment]
Hi @sumanthratna that's great! There is some support for svg writing [demoed in this example](https://github.com/napari/napari/blob/master/examples/to_svg.py) using the following parts of the code base  

https://github.com/napari/napari/blob/master/napari/layers/base/base.py#L658

https://github.com/napari/napari/blob/master/napari/layers/points/points.py#L1552

https://github.com/napari/napari/blob/1ae61bbb428ddf4eeef57aebc89bc6474bcb7751/napari/layers/shapes/shapes.py#L1380

but it was written a while ago and I don't know how widely used it is. I'd be curious if the existing functionality meets your needs? It might also need to be better documented, in which case contributions to our tutorials would be more than welcomed https://github.com/napari/tutorials

**Comment by tlambert03** [comment]
@sofroniewn and I just had a quick chat about writer plugins, here are the things we thought would be a good next step for #1104... keeping in mind that we will need to do slightly different things if the user wants to save a single layer vs multiple layers.

- there will be _both_ a hookspec as discussed above in #2, that accepts a list of layer types (as strings) and returns a writer function if it's capable of writing that combination of layers:
    ```python
    WriterFunction = Callable[[PathLike, List[LayerData], Optional[str]], None]

    @napari_hook_specification
    def napari_get_writer(
        path: str, layer_types: List[str], extension: Optional[str]
    ) -> Optional[WriterFunction]:
        ...
    ```
    as well as single hookspecs for each layer type:
    ```python
    @napari_hook_specification
    def napari_write_image(path: str, data: Any, metadata: dict, extension: str):
        ...

    ```
-  extension will be optional, and if provided will be appended to the filename, unless the filename already ends with the extension. (need to look into whether arguments to hook specs can be optional
- if a user has a single layer selected and hits "save", a dialog will appear that lets them choose a filename, and will also allow specifying which plugin to use (if multiple are available) based on plugins that implement a hookspec for the corresponding `napari_write_<layer_type>`, defaulting to the last used plugin.
- if multiple layers are selected when the user hits save, we call `napari_get_writer` and gather plugins that know how to handle the selected layers in combination.  (These may be further filtered by the filename or the extension), the user then selects an output path and the desired plugin as before.  (sidenote here: even though it should be fast to call, we can cache the response to `napari_get_writer` so we can quickly determine available plugins later, invalidating the cache whenever `plugin_manager.discover()` is called and returns a count > 0.)

**Comment by GenevieveBuckley** [comment]
If you're saving a list of layers (say a mixture of Images and Points), will this work as long as you have a writer that supports images  and a writer that supports points? Or do you need to have a single writer that supports both? If it's the latter, I can see that putting more of a burden on plugin authors.

**Comment by tlambert03** [comment]
We can make that work.  Here, I was describing the latter: in which `napari_get_writer` is specifically looking for plugins that know how to combine layer types into a single output.  Yes, that adds burden on plugin authors, but it's a tricky balance between making it easy for them, and also empowering them, since we definitely also need to somehow support the case where plugins _want_ that burden (in order to take advantage of specific combinations of layers into one unified output format...)

But that's also why we thought it would be useful to have single-layer (e.g. `napari_write_image`) hooks, that _we_ can combine: just looping single layers at a time into the save-with-plugin hook and dumping the output into a folder.

If one plugin wanted to support both single layers and some combination of layers, it would be up to them to figure out how to limit code duplication... maybe:

```python

def mega_writer(path, layer_data, extension):
    # do what you need to combine multiple points and images
    # into a single output, or, if layer_data is just a single item
    # with one type, handle that.

@napari_hook_specification
def napari_get_writer(path, layer_types, extension):
    if set(layer_types) == {'image', 'points'}:
        return mega_writer

@napari_hook_specification
def napari_write_image(path, data, metadata, extension):
    mega_writer(path, [(data, metadata)], extension)

@napari_hook_specification
def napari_write_points(path, data, metadata, extension):
    mega_writer(path, [(data, metadata)], extension)

```

**Comment by tlambert03** [comment]
just some saving "scenarios" from a chat with @sofroniewn ... these aren't that different from https://github.com/napari/napari/issues/1068#issuecomment-609074518 ... but rephrased slightly so want to record them somewhere:

**User selects “Save layer(s)...”**
_from menu, or calls `viewer.save_selected()`  or some utility function like `napari.save_layers`_

- If a single layer is selected, we just directly call `plugin_manager.hook.napari_write_<layer>()` which will loop through implementations and stop when the first one returns a non-`None` result. (The order in which implementations are called can be changed in the usual way with the implementation sorter/disabler. (We could also maybe show the current order in the Qt "Save dialog" when they go to pick a path)

- If multiple layers are selected, we call `plugin_manager.hook.napari_get_writer()` and let the first one that knows how to handle the combination of layers go for it. I’m not totally sure how we should handle exceptions in this case. It’d be nice to “clean up” and try the next one, but I’m not sure how to know what files got written (maybe inspect.stack?) or maybe we just leave that up to plugins?

- If no plugins offer `napari_get_writer` for that combination of layers, perhaps what our _own_ `napari_get_writer` implementation looks like is 1) create a folder 2) call `napari_write_<layer>` for each layer selected, altering the path such that they are written to the folder.

**User selects “Save layer(s) with plugin…”**
_from menu, or calls `viewer.save_selected(plugin=...)`  or some utility function like `napari.save_layers(plugin=...)`_
- if a single layer is selected, we call the `napari_write_<layer_type>` for that plugin, and if it fails it fails.
- if multiple layers are selected, we call `napari_get_writer` for that plugin, and if it doesn’t return a `WriterFunction` we error, otherwise we call it and let it fail if it fails.

> I do find myself wondering whether we’re thinking about “fallback” behavior too much.  Maybe users _always_ want to know exactly which writer plugin they’re getting, or none at all?  It's conceivable that the pluggy "call-all-the-plugins" model makes no sense here, and maybe users will only ever use "Save with" option ... but i think we have the bases covered with the two options.

**Comment by sofroniewn** [comment]
I have made a writer plugin that can write 2D images and points layers to svg -  https://github.com/sofroniewn/napari-svg now on pip to be used with #1104. More info coming soon

**Comment by GenevieveBuckley** [comment]
> I do find myself wondering whether we’re thinking about “fallback” behavior too much. Maybe users always want to know exactly which writer plugin they’re getting, or none at all? It's conceivable that the pluggy "call-all-the-plugins" model makes no sense here, and maybe users will only ever use "Save with" option ... but i think we have the bases covered with the two options.

I feel like we're getting a little too focused on *how* the layer is saved with a plugin instead of *what you get* saved to disk. 

I totally understand why we're all focused on the how, since we're programmers and that's a very necessary thing to know to implement it. But I don't think that way of thinking maps as well as we'd like to the kind of interface the user will need/want, so it's worth thinking about that part too.

I think that probably users are going to want three main things:
1. Save with a default method
    * Very few (or no) kwargs needed, a good sensible default.
    * Should be able to tell what the default plugin is relatively easily (and maybe change the default?)
2. Save as `$SPECIFIC_TYPE`
    * I don't think this is quite the same thing as saving with a specific plugin. If all I care about is having my data saved with geoJSON format to disk, I'm not going to care exactly which plugin handled it. It's easy to know what output type you want, hard to remember which plugins is the right one to pick for that.
3. Save layer(s) with a specific plugin
     * Presumably this would happen only when you have a very specific output requirement, or if another plugin is doing a poor job of saving a particular format.

tl:dr - I think it would be good to give users the option to say "I want this layer saved as tif/geoJSON/whatever" without needing to worry about which specific writer carries out the process.

**Comment by tlambert03** [comment]
actually, I think all three of these things are handled here already!  While there's no Qt implementation yet (working on it), from an API perspective, here's how you would do these things in #1104 as it currently stands:

> 1. save with a default method:

`viewer.save_layers(path)` ... will save all current layers to `path`.  The sensible default is our own `napari_get_writer` implementation.  Can look at the pluginsorter to see what the default is.

`viewer.save_layers(path, selected=True)` will save only the selected layers.

> 2. Save as specific type:

If all you care about is the format of the output, you just specify the extension to the path

`viewer.save_layers('something.geoJSON')` ... and whatever the first plugin is that knows how to give you what you ask for does it.  it's up to plugin implementations to look at your extension and know if they can handle it.  if no plugins are available, you'll get a message saying so.

> 3. Save layer(s) with a specific plugin

just specify `plugin_name`:

`viewer.save_layers(path, plugin_name='name_of_plugin')` 

that seem good?  all of the conversation above was definitely "in the weeds"... but we've always wanted a very clean and minimal API entry point.

(see: [`Viewer.save_layers`](https://github.com/napari/napari/blob/3de8f8673cd8657251c5b4e097fcd5082b0301c7/napari/components/add_layers_mixin.py#L900))

**Comment by GenevieveBuckley** [comment]
> actually, I think all three of these things are handled here already! 

Excellent! 😄 
I've just been reading through the latest changes to https://github.com/napari/napari/pull/1104, it's looking really nice
